File: src/com/machfour/macros/validation/Validations.java
package com.machfour.macros.validation;

import com.machfour.macros.data.Column;
import com.machfour.macros.util.DateStamp;

import java.util.Map;

public final class Validations {
    public static final Validation REQUIRED_FIELD;
    public static final Validation POSITIVE;
    public static final Validation INTEGRAL;
    public static final Validation NUMERIC;
    public static final Validation DATE;
    public static final Validation UNIQUE;

    static {
        REQUIRED_FIELD = new RequiredField();
        POSITIVE = new Positive();
        INTEGRAL = new Integral();
        NUMERIC = new Numeric();
        DATE = new Date();
        UNIQUE = new Unique();
    }

    private Validations() {
    }

    private static class RequiredField implements Validation {
        @Override
        public boolean validate(Map<Column, String> stringValues, Column field) {
            return !("".equals(stringValues.get(field)));
        }
    }

    private static class Numeric implements Validation {
        @Override
        public boolean validate(Map<Column, String> stringValues, Column field) {
            try {
                Double.valueOf(stringValues.get(field));
                return true;
            } catch (NumberFormatException e) {
                return false;
            }
        }
    }

    private static class Positive implements Validation {
        @Override
        public boolean validate(Map<Column, String> stringValues, Column field) {
            return NUMERIC.validate(stringValues, field) &&
                Double.valueOf(stringValues.get(field)) > 0;
        }

    }

    private static class Date implements Validation {
        @Override
        public boolean validate(Map<Column, String> stringValues, Column field) {
            try {
                DateStamp.fromIso8601String(stringValues.get(field));
                return true;
            } catch (IllegalArgumentException e) {
                return false;
            }
        }
    }

    private static class Integral implements Validation {
        @Override
        public boolean validate(Map<Column, String> stringValues, Column field) {
            try {
                Long.valueOf(stringValues.get(field));
                return true;
            } catch (NumberFormatException e) {
                // allow blank strings, since those should be validated by REQUIRED_FIELD
                return "".equals(stringValues.get(field));
            }
        }

    }

    private static class Unique implements Validation {
        @Override
        public boolean validate(Map<Column, String> stringValues, Column field) {
            //TODO
            return true;
        }
    }

}

File: src/main/java/com/machfour/macros/validation/Validations.kt
package com.machfour.macros.validation

import com.machfour.macros.core.Column
import com.machfour.macros.util.DateStamp

object Validations {
    val REQUIRED_FIELD = RequiredField()
    val POSITIVE = Positive()
    val INTEGRAL = Integral()
    val NUMERIC = Numeric()
    val DATE = Date()
    val UNIQUE = Unique()

    class RequiredField : Validation {
        override fun <M> validate(stringValues: Map<Column<M, *>, String?>, field: Column<M, *>): Boolean {
            return "" != stringValues[field]
        }
    }

    class Numeric : Validation {
        override fun <M> validate(stringValues: Map<Column<M, *>, String?>, field: Column<M, *>): Boolean {
            return stringValues[field]?.toDoubleOrNull() != null
        }
    }

    class Positive : Validation {
        override fun <M> validate(stringValues: Map<Column<M, *>, String?>, field: Column<M, *>): Boolean {
            return NUMERIC.validate(stringValues, field) && java.lang.Double.valueOf(stringValues[field]) > 0
        }
    }

    class Date : Validation {
        override fun <M> validate(stringValues: Map<Column<M, *>, String?>, field: Column<M, *>): Boolean {
            val dateString = stringValues[field] ?: return false;
            return try {
                DateStamp.fromIso8601String(dateString)
                true
            } catch (e: IllegalArgumentException) {
                false
            }
        }
    }

    class Integral : Validation {
        override fun <M> validate(stringValues: Map<Column<M, *>, String?>, field: Column<M, *>): Boolean {
            return try {
                java.lang.Long.valueOf(stringValues[field])
                true
            } catch (e: NumberFormatException) {
                // allow blank strings, since those should be validated by REQUIRED_FIELD
                "" == stringValues[field]
            }
        }
    }

    class Unique : Validation {
        override fun <M> validate(stringValues: Map<Column<M, *>, String?>, field: Column<M, *>): Boolean {
            //TODO
            return true
        }
    }

}